Flaky E2E: revamp the `SiteAssemberFlow` POM and the corresponding spec.

worldomonation opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Fixes #75812.

## Proposed Changes

This PR reworks the POM and the associated spec to hopefully eliminate an issue with the `Select Footer` step.

Key changes:
- merge the `selectHeader` and `selectFooter` methods into one.
- save and use the locator to perform `expect` checks.

## Testing Instructions

Ensure the following build configurations are passing:
  - [x] Pre-Release E2E

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

worldomonation committed:
packages/calypso-e2e/src/lib/flows/site-assembler-flow.ts
- simplify and combine methods into one.

test/e2e/specs/onboarding/setup__site-assembler.ts
- simplify the expect statements.
- reduce timeout for `site-editor` endpoint loading from 180 to 45 seconds.

worldomonation committed:
test/e2e/specs/onboarding/setup__site-assembler.ts
- remove comments

worldomonation committed:
test/e2e/specs/onboarding/setup__site-assembler.ts
- move definition

worldomonation commented on packages/calypso-e2e/src/lib/flows/site-assembler-flow.ts:
```diff
@@ -26,35 +22,20 @@ export class SiteAssemblerFlow {
-	}
+	async selectLayoutComponent( type: LayoutType, index = 0 ): Promise< void > {
+		await this.page.getByRole( 'button', { name: type } ).click();
+		await this.page.waitForLoadState( 'networkidle' );
```
While we are moving away from using the `networkidle` as indicators of loaded-ness, the flip side of the coin is that the site assembler flow fires off multiple requests to the same overall endpoint differing only in the parameters that are sent:

```
endpoint&param
endpoint&param&anotherparam
endpoint&yetanotherparam
endpoint&finalvariation
```

Normally, we'd wait on a specific response from an endpoint, but this behavior makes waiting on the completion of requests from the endpoint unpredictable, which is why I've resorted to using `networkidle`.

okmttdhr replied:
Thank you for addressing it!

FYI @Automattic/lego 
We might be able to reduce the requests somehow?

miksansegundo replied:
When choosing headers or footers, we should not fetch rendered patterns. That's a bug that is fixed in https://github.com/Automattic/wp-calypso/pull/75861.

karenroldan reviewed (approved):
The changes look good to me! Thanks for improving the tests. 🚀 

**note:** I ran this test 10x locally and got 2 failures out of 10. The test has some flakiness while waiting for `networkIdle` after clicking the `Header` layout type. But, we can monitor it in the meantime! The current implementation looks more stable. 👍🏻

okmttdhr commented on packages/calypso-e2e/src/lib/flows/site-assembler-flow.ts:
```diff
@@ -26,35 +22,20 @@ export class SiteAssemblerFlow {
-	 * Click "Continue" and land on the Site Editor
-	 */
-	async gotoSiteEditor(): Promise< void > {
-		// Wait for the "Continue" button to be enabled.
-		// @see: https://github.com/Automattic/wp-calypso/pull/75606
-		const waitFor = ( ms: number ) => new Promise( ( resolve ) => setTimeout( resolve, ms ) );
-		await waitFor( 500 );
```
https://github.com/Automattic/wp-calypso/pull/75807 made this change; it is to wait for the "Continue" button to be enabled, as its description says.
In this PR, we seem to be using `await page.waitForLoadState( 'networkidle' );` instead, I'd like to know if it ensures the button gets enabled. 👀

worldomonation replied:
@okmttdhr Yes, I'm sorry that my changes seem to undo a lot of what you did in #75807. I should probably explain why first - sorry for that.

Although we don't really have anything written down on this matter, past convention[[1](https://github.com/Automattic/wp-calypso/issues/57513#issuecomment-956540650)] as well as Playwright's own guidance on timeouts [[1](https://playwright.dev/docs/api/class-page#page-wait-for-timeout)] has led to a general avoidance of arbitrary timeouts in framework/POM code. 

That hasn't stopped me from trying to see if an arbitrary timeout _could_ solve problems elsewhere though, so I completely understand why it may look tempting. 

Given @karenroldan's observation that it still seems to fail 2/10 times (an unacceptably high incidence rate), I will take a second look to see if we can work around the weird button load issue in the specs. But if not, it may be required to have the feature team look at the button loading behavior. How does that sound?

okmttdhr replied:
@worldomonation 
You're right, generally speaking, it's better to avoid using arbitrary timeouts, which cause flaky tests, longer running time, etc, so I have tried the code below, but it did not work well for some reason. So I used "waitFor" (which is not ideal) 🙄 
(In case you missed it, disabling the button in 300ms is done intentionally for UX reasons. https://github.com/Automattic/wp-calypso/pull/75606)

```ts
  const locator = page.locator('#selector');
  await page.waitForFunction(([locator]) => {
    const element = locator.asElement()._element;
    return element && !element.disabled;
  }, locator);
```

> I will take a second look to see if we can work around the weird button load issue in the specs. But if not, it may be required to have the feature team look at the button loading behavior

Sounds good 👍  Thank you for all your cooperation!

worldomonation replied:
> disabling the button in 300ms is done intentionally for UX reasons. https://github.com/Automattic/wp-calypso/pull/75606

@okmttdhr Interesting, I was not aware of this reason. It means that we have to check for not just the presence of the `Continue` button but also for it to be active and enabled.

I think I have a possible answer, will iterate on this today.

worldomonation replied:
After https://github.com/Automattic/wp-calypso/pull/75865 went live, I was able to remove the `waitForTimeout` and leave just the operation to click on the `Continue` button and after 10 iterations locally, it all seems to work fine!

I'll run another 30 iterations, but I think this should work!

worldomonation commented:
> The changes look good to me! Thanks for improving the tests. 🚀
> 
> **note:** I ran this test 10x locally and got 2 failures out of 10. The test has some flakiness while waiting for `networkIdle` after clicking the `Header` layout type. But, we can monitor it in the meantime! The current implementation looks more stable. 👍🏻

@karenroldan Thanks for trying it locally and reporting back with the results. 2/10 times on local machine is unacceptably high, so I'll bring this back to the drawing board.

worldomonation committed:
packages/calypso-e2e/src/lib/flows/site-assembler-flow.ts
- split the method to action on the component type header and the actual method to select the component.

test/e2e/specs/onboarding/setup__site-assembler.ts
- wrap the final "Continue" button inside a Promise.all pattern that waits on the `/process/` URL.
- add comment explicitly warning that `waitForTimeout` is a temporary solution.

worldomonation commented:
Waiting for https://github.com/Automattic/wp-calypso/pull/75606#issuecomment-1512538989 to be resolved before pushing a new revision. 

Once the `disabled` attribute is added to the `Continue` button for 300ms, it _should_ stop tripping up Playwright.

worldomonation committed:
test/e2e/specs/onboarding/setup__site-assembler.ts
- remove the temporary `watitForTimeout`

okmttdhr reviewed (approved):
LGTM 👍 
Thank you for being so cooperative!
